cmake/modules/zephyr_default.cmake

```diff
@@ -81,9 +82,43 @@ include(${BOARD_DIR}/pre_dt_board.cmake OPTIONAL)
 # preprocess DT sources, and then, after we have finished processing
 # both DT and Kconfig we complete the target-specific configuration,
 # and possibly change the toolchain.
-include(generic_toolchain)
-include(dts)
-include(kconfig)
-include(soc)
-include(target_toolchain)
-include(kernel)
+list(APPEND sub_components generic_toolchain)
+list(APPEND sub_components dts)
+list(APPEND sub_components kconfig)
+list(APPEND sub_components soc)
+list(APPEND sub_components target_toolchain)
+
+set(load_module_index -2)
+foreach(component ${SUB_COMPONENTS})
+  list(FIND sub_components ${component} current_module_index)
+  list(FIND sub_components "\${${component}}" escaped_module_index)
+  if(${current_module_index} EQUAL -1 AND ${escaped_module_index} EQUAL -1)
+    message(FATAL_ERROR
+      "Subcomponent ${module} not default module for Zephyr CMake build system.\n"
+      "Please choose one or more valid components: ${sub_components}"
+    )
+  endif()
+
+  if(${current_module_index} GREATER ${load_module_index})
+    set(load_module_index ${current_module_index})
+  elseif(${escaped_module_index} GREATER ${load_module_index})
+    set(load_module_index ${escaped_module_index})
+  endif()
+endforeach()
+
+math(EXPR load_module_index "${load_module_index} + 1")
+list(SUBLIST sub_components 0 ${load_module_index} load_modules)
+
+foreach(module IN LISTS load_modules)
+  # Ensures any module of type `${module}` are properly expanded to list before
+  # passed on the `include(${module})`.
+  # This is done twice to support cases where the content of `${module}` itself
+  # contains a variable, like `${BOARD_DIR}`.
+  string(CONFIGURE "${module}" module)
+  string(CONFIGURE "${module}" module)
+  include(${module})
+endforeach()
+
+if(NOT DEFINED SUB_COMPONENTS)
+  include(kernel)
+endif()
```

Comment:
Aren't we being overly-generic here?

Isn't it much simpler and easier for people to understand this code if we did

if(STOP_AFTER_DT)
return()
endif()


if(STOP_AFTER_KCONF)
return()
endif()

Reply:
> Aren't we being overly-generic here?
> 
> Isn't it much simpler and easier for people to understand this code if we did

For a maintainer of this code, I partly agree.
But not for the users of the tool.

I don't expect users of this tool to have internal knowledge of the order in which Zephyr CMake modules are loaded.

So if a user or twister needs to have DTS and kconfig loaded they should not need to know which part is loaded first, they can simply say, *I want dts and kconfig* by invoking:
```
$ cmake -DBOARD=<board> -S samples/hello_world -DMODULES="kconfig;dts" -P cmake/package_tooling.cmake
```

and the script will ensure that both dts and kconfig are loaded.

With your proposal, a user / tool will need to consider, *I want both dts and kconfig, is dts executed loaded first, or is kconfig ?*, so do I need to do: `-DSTOP_AFTER_KCONF=true` or `-DSTOP_AFTER_DT=true`.

Reply:
not to mention, the current proposal ensures the Zephyr CMake module names are used, and not a derived setting name.

If user makes a typo or otherwise writes an invalid module name, for example: `-DMODULES=dtts` an error like this is printed:
```
Error: Subcomponent dtts not default module for Zephyr CMake build system.
         Please choose one or more valid components: <whole-list-of-valid-modules>"
```

whereas with your proposal, users need to know the relation between the module name and corresponding CMake variable, like: `dts` module corresponds to `STOP_AFTER_DT`.

And if writing `-DSTOP_AFTER_DTS=true` instead, then everything is processed and no error message is returned to the user.

Reply:
IMHO it is reasonable to expect that anyone that wants to halt execution after both DT and Kconfig have executed knows that DT executes before Kconfig.

IMHO it is also acceptable that obscure errors happen if the user uses the wrong module name.

This niche feature just isn't worth all this complexity we are adding.

Reply:
> This niche feature just isn't worth all this complexity we are adding.

we are adding Zephyr CMake modules to a list, and then looping that list until all modules requested by user has been loaded.

How can adding to a list and looping that list be considered complex ?

Reply:
To me it is very clear that adding two return statements and two if statements will be much simpler
for the user to understand than the patch we have now. I'm not sure how to argue the point.
It's just a matter of the amount of code, logic, and abstractions, needed to get the job done.

That being said, if you insist on this approach then we should at least reduce the compexity as much as we can.

e.g. we can create a function:

```
# write to sorted_sub_list_output, a sublist of sorted_list_input starting from the beginning of the list that contains
# all items in list_input
find_sublist_with_all(sorted_list_input list_input sorted_sub_list_output)
```

to remove some of the noise from the flow of the program.

Reply:
> To me it is very clear that adding two return statements and two if statements will be much simpler for the user to understand than the patch we have now.

Aah, so you think we should support only DTS and Kconfig.
I see this as a general way to allow us to run just run up to a certain point.
Other use-cases could be `zephyr_modules` or toolchain handling.

For example, we might be able to clean up this in future (outside the scope of this PR):
https://github.com/zephyrproject-rtos/zephyr/blob/8751a0feb0388698698197f08fbf5399917650c2/scripts/pylib/twister/twisterlib.py#L3029-L3030

in that case, you would need to add a CMake variable and check at each possible return point someone might use, not just two if and two return statements.

Reply:
complexity is now acceptable after the cleanup.